src/mame/drivers/jpmsru.cpp

```diff
@@ -59,6 +64,97 @@
 
 #define MAIN_CLOCK 6_MHz_XTAL
 
+class jpmsru_state : public genfruit_class
+{
+public:
+	jpmsru_state(const machine_config &mconfig, device_type type, const char *tag) : 
+			genfruit_class(mconfig, type, tag),
+			m_maincpu(*this, "maincpu"),
+			m_inputs(*this, "IN%u", 0U),
+			m_reel(*this, "reel%u", 0U),
+			m_lamp(*this, "lamp%u", 0U),
+			m_digits(*this, "digit%u", 0U),
+			m_audio_in(*this, "nl_audio:in%u", 0U),
+			m_nvram(*this, "nvram", 0x80, ENDIANNESS_BIG),
+			m_dips(*this, "DIP%u", 0U)
+	{ }
+
+	void jpmsru_3k(machine_config &config);
+	void jpmsru_3k_busext(machine_config &config);
+	void jpmsru_4k(machine_config &config);
+	void ewn(machine_config &config);
+	void ewn2(machine_config &config);
+	void ndu(machine_config &config);
+	void dud(machine_config &config);
+	void lan(machine_config &config);
+
+	void init_jpmsru();
+	
+	template <unsigned N> DECLARE_READ_LINE_MEMBER(opto_r) { return m_opto[N]; }
+protected:
+	virtual void machine_start() override;
+
+private:
+	template <unsigned N> DECLARE_WRITE_LINE_MEMBER(opto_cb) { m_opto[N] = state; }
+
+	uint8_t inputs_r(offs_t offset);
+	void reel_w(offs_t offset, uint8_t data);
+	void update_int();
+	void audio_w(offs_t offset, uint8_t data);
+	void int1_en_w(offs_t offset, uint8_t data);
+	void int2_en_w(offs_t offset, uint8_t data);
+	uint8_t busext_data_r(offs_t offset);
+	void busext_data_w(offs_t offset, uint8_t data);
+	void busext_bdir_w(offs_t offset, uint8_t data);
+	void busext_mode_w(offs_t offset, uint8_t data);
+	void busext_addr_w(offs_t offset, uint8_t data);
+	uint8_t busext_dips_r(offs_t offset);
+	void out_lamp_w(offs_t offset, uint8_t data);
+	void out_lamp_ext_w(offs_t offset, uint8_t data);
+	void out_disp_w(offs_t offset, uint8_t data);
+	void out_payout_cash_w(offs_t offset, uint8_t data);
+	void out_payout_token_w(offs_t offset, uint8_t data);
+	template<unsigned meter> void out_meter_w(offs_t offset, uint8_t data);
```

Comment:
Please use `LlamaCase` for template parameter names to make it clearer that they’re semantically different to function parameters.

Reply:
Fixed.